From 205d40b90156f6f47ee09835488d1ab4a1809fcb Mon Sep 17 00:00:00 2001
From: Stephan Hartmann <stha09@googlemail.com>
Date: Thu, 4 May 2023 13:48:13 +0200
Subject: fix crash in InkDropHost

On destruction of InkDropHost, InkDropImpl gets destructed as well, but calls RemoveInkDropLayer of partly destructed InkDropHost object. This
method resets ink_drop_mask_ which is already destructed. To avoid
this, reset ink_drop_ early.

Bug: 1434712
Change-Id: Id3c5dbb70aa18328a52f024b5d2fa2be129efa9a
---
 ui/views/animation/ink_drop_host.cc | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/ui/views/animation/ink_drop_host.cc b/ui/views/animation/ink_drop_host.cc
index 0255bce16ecd7..76a4204f05353 100644
--- a/ui/views/animation/ink_drop_host.cc
+++ b/ui/views/animation/ink_drop_host.cc
@@ -74,7 +74,12 @@ InkDropHost::InkDropHost(View* view)
       ink_drop_event_handler_delegate_(this),
       ink_drop_event_handler_(view, &ink_drop_event_handler_delegate_) {}
 
-InkDropHost::~InkDropHost() = default;
+InkDropHost::~InkDropHost() {
+  // Destroy ink_drop_ early as it holds references to this
+  // and should not call methods on a partially-destructed
+  // InkDropHost as part of its destruction.
+  ink_drop_.reset();
+}
 
 std::unique_ptr<InkDrop> InkDropHost::CreateInkDrop() {
   if (create_ink_drop_callback_) {
-- 
2.39.2

